Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/SLList.h

```diff
@@ -0,0 +1,374 @@
+	//											SIZE
+	//-------------------------------------------------------------------------------------------
+	bool IsEmpty() const
+	{
+		return Size == 0;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	int GetSize() const
+	{
+		return Size;
+	}
+
+
+	//-------------------------------------------------------------------------------------------
+	//										MODIFIERS
+	//-------------------------------------------------------------------------------------------
+	void AddHead(const T& NewElement)
+	{
+		FNode* Node = new FNode;
+		Node->Element = NewElement;
+		Node->Next = Head;
+		Head = Node;
+
+		if (!Tail)
+		{
+			Tail = Head;
+		}
+		Size++;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void AddTail(const T& NewElement)
+	{
+		if (Tail)
+		{
+			FNode* Node = new FNode;
+			Node->Element = NewElement;
+			Node->Next = nullptr;
+
+			Tail->Next = Node;
+			Tail = Node;
+
+			Size++;
+		}
+		else //If Tail doesn't exist, there is no Head either
+		{
+			AddHead(NewElement);
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Insert(const T& NewElement, const int Index)
+	{
+		if (Index < 0 || Index > Size) //Bounds Checking
+		{
+			return;
+		}
+		else if (Index == 0) //Head
+		{
+			AddHead(NewElement);
+		}
+		else if (Index == Size)
+		{
+			AddTail(NewElement);
+		}
+		else if(Index < Size)
+		{
+			FNode* Previous = GetNode(Index-1);
+
+			FNode* NewNode = new FNode();
+			NewNode->Element = NewElement;
+			NewNode->Next = Previous->Next;
+
+			Previous->Next = NewNode;
+
+			Size++;
+		}	
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Remove(const int Index)
+	{
+		if (Index < 0 || Index >= Size) //Bounds Checking
+		{
+			return;
+		}
+		else if (Index == 0) //Head
+		{
+			FNode* NewHead = Head->Next;
+			delete Head;
+			Head = NewHead;
+
+			if (Size == 1)
+			{
+				Tail = nullptr;
+			}
+		}
+		else
+		{
+			FNode* Previous = GetNode(Index - 1);
+			FNode* Current = Previous->Next;
+			Previous->Next = Current->Next;
+			delete Current;
+
+			if (Index == Size - 1)
+			{
+				Tail = Previous;
+			}
+		}
+		Size--;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Clear()
+	{
+		for (int i = 0; i < Size; i++)
+		{
+			FNode* Current = Head;
+			Head = Head->Next;
+			delete Current;
+		}
+		Tail = nullptr;
+		Size = 0;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	//								TEMPLATES & PREDICATES
+	//-------------------------------------------------------------------------------------------
+	template<typename Pred>
+	void ForEach(const Pred& Predicate)
+	{
+		FNode* Current = Head;
+		for (int i = 0; i < Size; i++)
+		{
```

Comment:
you could already implement the iterators with this function something like 

`for (T& Temporal : *this)
		{
			if (Predictate(Temporal))
			{
				return &Temporal->Element;
			}
		}`